S3Outputer.cc

```diff
@@ -0,0 +1,230 @@
+}
+void S3Outputer::output(
+    EventIdentifier const& iEventID,
+    SerializeStrategy const& iSerializers,
+    TaskHolder iCallback
+    ) const
+{
+  using namespace std::string_literals;
+  if(verbose_ >= 2) {
+    std::cout <<"   run:"s+std::to_string(iEventID.run)+" lumi:"s+std::to_string(iEventID.lumi)+" event:"s+std::to_string(iEventID.event)+"\n"<<std::flush;
+  }
+  auto sev = currentEventStripe_.add_events();
+  sev->set_offset(eventGlobalOffset_++);
+  sev->set_run(iEventID.run);
+  sev->set_lumi(iEventID.lumi);
+  sev->set_event(iEventID.event);
+
+  auto p = std::begin(currentProductStripes_);
+  auto pi = index_.mutable_products()->begin();
+  for(const auto& s : iSerializers) {
+    if (verbose_ >= 2) {
+      std::cout << "adding blob len " << s.blob().size() << " to " << pi->productname() << "\n";
+      for (auto c : s.blob()) {
+        if ( isprint(c) ) std::cout << c;
+        else std::cout << "\\x" << std::hex << (int) c << std::dec;
+      }
+      std::cout << "\n";
+    }
+    p->mutable_content()->append(s.blob().begin(), s.blob().end());
+    p->add_offsets(p->content().size());
+    p++; pi++;
+  }
+
+  flushProductStripes(iCallback);
+
+  if ( currentEventStripe_.events_size() == eventFlushSize_ ) {
+    if(verbose_ >= 2) {
+      std::cout << "reached event flush size "s + std::to_string(eventFlushSize_) + ", flushing\n" << std::flush;
+    }
+    flushEventStripe(iCallback);
+  }
+}
+
+
+void S3Outputer::flushProductStripes(TaskHolder iCallback, bool last) const {
+  using namespace std::string_literals;
+  auto p = currentProductStripes_.begin();
+  auto pi = index_.mutable_products()->begin();
+  for(; p != std::end(currentProductStripes_); ++p, ++pi) {
+    size_t offset = p->content().size();
+    size_t bufferNevents = p->offsets_size();
+
+    // first flush when we exceed min size and have an even divisor of eventFlushSize_
+    // subsequent flush when we reach productFlushSize
+    // always flush when we reach eventFlushSize_ (for buffers that never get big enough)
+    // flush if last call and we have something to write
+    if (
+        ((pi->flushsize() == 0) && (offset > productBufferFlushMinSize_) && (eventFlushSize_ % bufferNevents == 0))
+        || (bufferNevents == pi->flushsize())
+        || (bufferNevents == eventFlushSize_)
+        || (last && bufferNevents > 0)
+        )
+    {
+      if(verbose_ >= 2) {
+        std::cout << "product buffer for "s + std::string(pi->productname()) + " is full ("s + std::to_string(offset)
+          + " bytes, "s + std::to_string(bufferNevents) + " events), flushing\n" << std::flush;
+      }
+      objstripe::ProductStripe pOut;
+      pOut.mutable_offsets()->Reserve(bufferNevents);
+      pOut.mutable_content()->reserve(offset);
+      std::swap(*p, pOut);
+      std::string name = objPrefix_;
+      name += pi->productname();
+      name += std::to_string(eventGlobalOffset_ - bufferNevents);
+      iCallback.group()->run(
```

Comment:
Once you startup a new group-run, you are no longer guaranteed serialization since you just created a new task. This task could get stalled and one for the same product could be created and started at the same time as this one. It seems to me that this doesn't work as you intend.

Reply:
This swaps the output buffer with a fresh one (for the next data products) and writes the old buffer out to a key in the object store, so the ordering of executing the lambda here is not important. But maybe I misunderstand, `group()->run(...)` is non-blocking right? So the rest of this function continues in the current thread and the new task will just run when it can?

Reply:
task_group::run is non-blocking and runs either on this thread when all other tasks queued on this thread ahead of it are done or on another thread if that thread has no queued work for itself.